stdlib/public/core/ManagedBuffer.swift

```diff
@@ -136,6 +136,38 @@ open class ManagedBuffer<Header, Element> {
   }
 }
 
+public extension ManagedBuffer {
+  @inline(never)
+  final class func reallocate<Buffer: ManagedBuffer>(
+    buffer: inout Buffer,
+    count: Int,
+    newMinimumCapacity: Int
+    ) {
+      precondition(_isBitwiseTakable(Header.self))
+      precondition(_isBitwiseTakable(Element.self))
+      precondition(isKnownUniquelyReferenced(&buffer))
+
+      let newSizeInBytes = MemoryLayout<Header>.stride
+        + newMinimumCapacity * MemoryLayout<Element>.stride
+
+      // ref count: 1
+      let oldBufferPtr = Unmanaged.passRetained(buffer).toOpaque()
+      // ref count: 2
+      let newBufferPtr = _stdlib_realloc(oldBufferPtr, newSizeInBytes)
+      // ref counts: realloc worked: 2 (old ptr is new ptr)
+      //             realloc failed: old ptr: 2, new ptr: 2
+      buffer = Unmanaged.fromOpaque(newBufferPtr).takeRetainedValue()
```

Comment:
so I think this is just wrong in the case that `realloc` fails (ie. old and new buffer have _different_ addresses): assigning to `buffer` will then decrement the reference count old the old `buffer` (which is already free'd  🔥) and increment the reference count of the new buffer (okay but unnecessary).

Reply:
Given that ManagedBuffer is a Swift class, you shouldn't need (or want) to do any refcounting. Instead, treat the old and new pointers as opaque pointers and assign like that. You may want to take `buffer` as an `UnsafeMutablePointer<Buffer>` to have more control over the refcounting. You could probably do it with `withUnsafeMutablePointer(to:)` but taking the parameter as a UMP seems like less magic.

Reply:
@mikeash Thanks, that sounds like a good plan!

Regarding the side table: Yes, so the initial idea was to change `isKnownUniquelyReferenced` to only return `true` iff refcount == 1 && no side table. I filed [SR-8538](https://bugs.swift.org/browse/SR-8538) for that. But then @lorentey previously had a use-case for exactly the opposite behaviour [SR-5633](https://bugs.swift.org/browse/SR-5633). Someone then proposed to deprecate `isKnownUniquelyReferenced` and introduce two new functions with better names that make it clear what happens with weak/unowned references.
I'm happy to file an SR about that if you think that's a good plan.

For now (ie. this patch) I think it would be good enough to precondition that there's no side-table. Is there a way to do that already?

Reply:
`RefCount.h` has a `hasSideTable` method, but I don't see anywhere that exposes it to Swift. Your best bet there is probably to add a new runtime function that calls through to that. `isKnownUniquelyReferencedSeriouslyEvenIncludingTheSideTableYesIMeanIt`